### DISCUSSION TITLE
Use Eloquent accessor for recordTitleAttribute in Relationship manager

### DISCUSSION THREAD

**Discussion by rikiless** [filamentphp/filament#5302] [Help]
Hello

Is it possible to use Eloquent accessor as parameter for `recordTitleAttribute` in Relationship Manager?

According to [docs](https://filamentphp.com/docs/2.x/admin/resources/getting-started#record-titles) it should work.

> You may specify the name of an [Eloquent accessor](https://laravel.com/docs/eloquent-mutators#defining-an-accessor) if just one column is inadequate at identifying a record.

```php

 public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make('fullName'),
            ])
            ->filters([
                //
            ])
            ->headerActions([
                Tables\Actions\CreateAction::make(),
                Tables\Actions\AssociateAction::make()
                    ->recordTitleAttribute('fullName') // <-------- There's problem.
                    ->preloadRecordSelect(),
// ...
```

It will throws following error:

<img width="2560" alt="Screenshot 2022-12-21 at 12 59 28" src="https://user-images.githubusercontent.com/589851/208900859-d4009463-2f4d-44df-8381-c5d43f49ab15.png">

The same isue if i use following in RelationShipManager class:

```php
protected static ?string $recordTitleAttribute = 'fullName';
```

Accessor in `User` model is defined this way:

```php
public function fullName(): Attribute
    {
        return Attribute::make(
            get: fn ($value, $attributes) => trim("{$attributes['name']} {$attributes['surname']}"),
        );
    }
```

Thanks.

**Comment by ArtMin96** [comment]
```php
protected static ?string $recordTitleAttribute = 'fullName';
```

to

```php
protected static ?string $recordTitleAttribute = 'full_name';
```

**Reply by rikiless** [reply]
The same issue :( 

> SQLSTATE[42S22]: Column not found: 1054 Unknown column 'full_name' in 'order clause'

In this resource I also have using following and it works okay.
```php
public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make('full_name'), // fullName is also ok
                // ...
                ```

**Reply by rikiless** [reply]
So maybe it's broken with use of `Tables\Actions\AssociateAction::make()`?

**Reply by ArtMin96** [reply]
replace public with protected as mentioned in the doc։ [https://laravel.com/docs/9.x/eloquent-mutators#defining-an-accessor](https://laravel.com/docs/9.x/eloquent-mutators#defining-an-accessor)

```php
protected function fullName(): Attribute
```

and then `full_name`

**Reply by rikiless** [reply]
I fixed it according to docs, but it did not help. The same error :(

**Comment by danharrin** [comment]
Try using a virtual column in the database instead of an accessor.

```php
$table->string('full_name')->virtualAs('concat(first_name, \' \', last_name)');
```

**Reply by rikiless** [reply]
Nice. I will use this as workaround for now. Thanks.

**Reply by danharrin** [reply]
Alternatively, try `AssociateAction::make()->recordSelect(fn ($select) => $select->searchColumns(['first_name', 'last_name']))`

**Reply by rikiless** [reply]
It returned `Method Filament\Forms\Components\Select::searchColumns does not exist.` But I'm fine with previous solution (`virtualAs`). I also use `preloadRecordSelect()`.

Shouldn't be original issue be reported as bug? Using Eloquent accessor it works in table listing, but not for `AssociateAction` in header action.

**Reply by danharrin** [reply]
No its not a bug, there's no way we can fix it as SQL doesnt understand Laravel accessors

**Reply by cainydev** [reply]
Possibly dont query database with `where` clause if recordTitle is an accessor and apply filtering afterwards? (Global Search is fucked up with accessor)

**Reply by danharrin** [reply]
Good luck filtering over 500 records without SQL 😁

**Reply by cainydev** [reply]
Thats true but it works if its not too much. Problem with virtual columns is that they cant access related models. So if I have a model which contains only numbers or foreign ids, i dont have much choice as to what i make searchable. If you apply the search functions (getGlobalSearchResultTitle) the filtering is still done in php..

is there any support for elastic search or similars? I couldnt find anything in the docs

**Reply by danharrin** [reply]
If you want to search related records, you can use whereHas() or a join?

**Comment by pgicharu** [comment]
I faced the same challenge and came across this discussion. Here's what worked for me: In the resource page I made use of the `getGlobalSearchResultTitle` and `getGloballySearchableAttributes` methods to achieve the desired results like so:

```php
    public static function getGlobalSearchResultTitle(\Illuminate\Database\Eloquent\Model $record): string
    {
        return $record->full_name; //fullName works too
    }

    public static function getGloballySearchableAttributes(): array
    {
        return ['first_name', 'last_name'];
    }
```

**Comment by neapolis79** [comment]
for me worked use 

`->recordTitle(fn ($record) => $record->full_name)`

instead of 

`->recordTitleAttribute('full_name')`

like indicated in this issue #9566
